### DISCUSSION TITLE
Direct serial connection... Remotely?

### DISCUSSION THREAD

**Discussion by Karakth** [wlcrs/huawei_solar#95] [Setting up the integration]
Hey, 

First off thanks for all the hard work for this amazing project. 
Secondly, with the latest update comes a new recommended connection mode via direct serial connection. My issue is that the inverter is on the roof and my HA server is in the basement several floors away. Any ideas (short of running a usb cable for that distance) to connect it? Would it be possible to expose the sensors on a raspberry pi and communicating over MQTT?

**Comment by wlcrs** [comment]
[The main reason that I promote this solution as "recommended" is that now 95% of all issues/posts come from people that are having trouble setting up the networking part, and I even read about people buying routers just to set up the connection. For them it is benefitial to buy something that is simpler to setup. However, if you already have a working network setup there is no real benefit to change to the *direct serial* approach instead.

On to your question: I would look into a ModbusRTU (=serial) to ModbusTCP (=network) bridge.

If you already have a Pi lying around, you could look into using https://github.com/3cky/mbusd to set up such a bridge.

I already have an [M5 Atom RS485](https://shop.m5stack.com/products/atom-rs485-kit?ref=WkCovois) lying around that I want to flash with a [modbus bridge](https://emodbus.github.io/modbusbridge) to test a remote scenario as yours. But this is likely something that I only will get to during the winter months. There also exist commercial solutions like the [EW11](https://s.click.aliexpress.com/e/_DDcMOYb) (sometimes also marketed as HF7211-0) to create such a bridge, and I know of at least one person who is already successfully using this.

**Reply by JacoboSanchez** [reply]
I am the one with the EW11A device, we talked about it [here](https://github.com/wlcrs/huawei_solar/issues/21#issuecomment-1030713725). I have it on two different inverters. One of them since February and it is working fine . For the other one I was having some issues with powering it until I realiced some days ago that I can use the 5V from the dongle USB connector... (just need to figure out a way to isolate it properly to not have it exposed to humidity in the air)

**Comment by JuanmanDev** [comment]
Hello! 

Sorry for reopening this ropex 2 years later.

I have some problems sharing the serials connections directly from proxmox to a LXC with docker and exposing all so after 2 days i face a way to be able to expose the Huawei solar over networks on a different machine (is the same machine but different LXC, so I could migrate in the future).

First of all I need to share the usb serial to my LXC, but I follow this (you would not need this probably): 
https://forum.proxmox.com/threads/how-to-usb-passthrough-into-an-unpriviledged-lxc-with-the-ui-in-8-2.150390/


To convert the machine in a Modbus TCP server that have a Modbus RTU connected following the guide in this repo, you would need mbusd, in alpine linux you would have to download and compile.

Them you have to run every time the system boots or create a service like above.

Update the port path before run ;) 


```shell
#!/bin/sh

# Update package list
echo "Updating package list..."
apk update

# Step 1: Install required dependencies for building mbusd
echo "Installing build dependencies..."
apk add build-base cmake git libtool automake autoconf linux-headers pkgconfig openrc

# Step 2: Clone mbusd source code from GitHub
echo "Cloning mbusd repository..."
git clone https://github.com/3cky/mbusd.git

# Step 3: Build mbusd from source
echo "Building mbusd from source..."
cd mbusd
mkdir build
cd build
cmake ..
make

# Step 4: Install mbusd
echo "Installing mbusd..."
make install

# Step 5: Create OpenRC service script for mbusd
echo "Creating mbusd OpenRC service script..."

cat << 'EOF' > /etc/init.d/mbusd
#!/sbin/openrc-run

description="Modbus RTU to TCP service for solar inverter"

command="/usr/local/bin/mbusd"
command_args="-L /tmp/mbusd.log -p /dev/ttyUSB_SERIAL_SOLAR0 -s 9600 -P 502 -d -v9"
pidfile="/run/mbusd.pid"
command_background=true

depend() {
    need localmount
}
EOF

# Step 6: Make the service script executable
echo "Making mbusd service script executable..."
chmod +x /etc/init.d/mbusd

# Step 7: Add mbusd service to the default runlevel to start on boot
echo "Adding mbusd service to default runlevel..."
rc-update add mbusd default

# Step 8: Start the mbusd service
echo "Starting mbusd service..."
rc-service mbusd start

# Step 9: Verify service status
echo "Checking mbusd service status..."
rc-service mbusd status

echo "mbusd installation and setup is complete."

```

After this you would be able to create a remote connection using a raspbery pi + cheap modbus usb.


But if you prefer something more quick and easy, you could buy one of the cheap wifi modbus adapters: https://aliexpress.com/item/32918411760.html

![image](https://github.com/user-attachments/assets/a18230f9-924a-40db-be24-70665b81af1b)

Hope it helps to someone!
